// tspec/dist/cli.js
#!/usr/bin/env node
import { realpathSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import { d as defaultGenerateParams, g as generateTspec, i as initTspecServer } from './chunks/index.js';
import 'express';
import 'http-proxy-middleware';
import 'swagger-ui-express';
import 'fs/promises';
import 'path';
import 'debug';
import 'glob';
import 'typescript';
import 'typescript-json-schema';
import 'json-schema-to-openapi-schema';

var SupportedSpecVersion; (function (SupportedSpecVersion) {
  const THREE = 3; SupportedSpecVersion[SupportedSpecVersion["THREE"] = THREE] = "THREE";
})(SupportedSpecVersion || (SupportedSpecVersion = {}));



















const baseOptions = {
  specPathGlobs: { type: 'array', default: defaultGenerateParams.specPathGlobs },
  tsconfigPath: { type: 'string', default: defaultGenerateParams.tsconfigPath },
  configPath: { type: 'string', default: defaultGenerateParams.configPath },
  outputPath: { type: 'string' },
  specVersion: { type: 'number' },
  openapiTitle: { type: 'string', default: defaultGenerateParams.openapi.title },
  openapiVersion: { type: 'string', default: defaultGenerateParams.openapi.version },
  openapiDescription: { type: 'string', default: defaultGenerateParams.openapi.description },
  debug: { type: 'boolean', default: defaultGenerateParams.debug },
  ignoreErrors: { type: 'boolean', default: defaultGenerateParams.ignoreErrors },
} ;

const generatorOptions = {
  ...baseOptions,
  outputPath: { type: 'string', default: 'openapi.json' },
} ;

const runServerOptions = {
  ...baseOptions,
  port: { type: 'number', default: 7000 },
  proxyHost: { type: 'string' },
} ;

const validateGeneratorOptions = (args) => {
  if (args.specVersion && !Object.values(SupportedSpecVersion).includes(args.specVersion)) {
    // eslint-disable-next-line max-len
    throw new Error(`Tspec currently supports only OpenAPI Spec with version ${Object.values(SupportedSpecVersion).join(', ')}.`);
  }

  return {
    specPathGlobs: args.specPathGlobs !== defaultGenerateParams.specPathGlobs
      ? args.specPathGlobs.map((glob) => glob.toString())
      : undefined,
    tsconfigPath: args.tsconfigPath !== defaultGenerateParams.tsconfigPath ? args.tsconfigPath : undefined,
    configPath: args.configPath !== defaultGenerateParams.configPath ? args.configPath : undefined,
    outputPath: args.outputPath,
    specVersion: args.specVersion !== defaultGenerateParams.specVersion ? args.specVersion : undefined,
    openapi: {
      title: args.openapiTitle !== defaultGenerateParams.openapi.title ? args.openapiTitle : undefined,
      version: args.openapiVersion !== defaultGenerateParams.openapi.version ? args.openapiVersion : undefined,
      description: args.openapiDescription !== defaultGenerateParams.openapi.description
        ? args.openapiDescription
        : undefined,
    },
    debug: args.debug !== defaultGenerateParams.debug ? args.debug : undefined,
    ignoreErrors: args.ignoreErrors !== defaultGenerateParams.ignoreErrors ? args.ignoreErrors : undefined,
  };
};

const specGenerator = async (args) => {
  const generateTspecParams = await validateGeneratorOptions(args);
  await generateTspec(generateTspecParams);
};

const startTspecServer = async (args) => {
  const generateTspecParams = await validateGeneratorOptions(args);
  initTspecServer({ ...generateTspecParams, port: args.port, proxyHost: args.proxyHost });
};

const runCli = async () => yargs(hideBin(process.argv))
  .usage('Usage: $0 <command> [options]')
  .command(
    'generate',
    'Generate OpenAPI Spec from Tspec',
    generatorOptions,
    (yargs) => specGenerator(yargs),
  )
  .command(
    'server',
    'Start Tspec server',
    runServerOptions,
    (yargs) => startTspecServer(yargs),
  )
  .help('help')
  .alias('help', 'h')
  .parse();

if (import.meta.url.startsWith('file:')) {
  const modulePath = realpathSync(fileURLToPath(import.meta.url));
  if (realpathSync(process.argv[1]) === modulePath) {
    runCli();
  }
}

export { runCli };
